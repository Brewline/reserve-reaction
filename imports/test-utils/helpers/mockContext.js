const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
