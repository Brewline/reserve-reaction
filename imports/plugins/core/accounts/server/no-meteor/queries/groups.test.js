import mockContext from "/imports/test-utils/helpers/mockContext";
import groupsQuery from "./groups";

const fakeShopId = "FAKE_SHOP_ID";
const fakeAccount = { _id: "FAKE_ACCOUNT_ID", groups: ["group1", "group2"] };

beforeEach(() => {
  jest.resetAllMocks();
});

test("returns the groups cursor if userHasPermission returns true", async () => {
  mockContext.collections.Groups.find.mockReturnValueOnce("CURSOR");
  mockContext.userHasPermission.mockReturnValueOnce(true);
  const result = await groupsQuery(mockContext, fakeShopId);
  expect(mockContext.collections.Groups.find).toHaveBeenCalledWith({ shopId: fakeShopId });
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin", "reaction-accounts"], fakeShopId);
  expect(result).toBe("CURSOR");
});

test("returns the groups cursor for groups the current user is in, if userHasPermission returns false", async () => {
  mockContext.collections.Groups.find.mockReturnValueOnce("CURSOR");
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
  const result = await groupsQuery(mockContext, fakeShopId);
  expect(mockContext.collections.Groups.find).toHaveBeenCalledWith({
    _id: { $in: fakeAccount.groups },
    shopId: fakeShopId
  });
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin", "reaction-accounts"], fakeShopId);
  expect(result).toBe("CURSOR");
});

test("throws access-denied if not allowed", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
  const result = groupsQuery(mockContext, fakeShopId);
  return expect(result).rejects.toThrowErrorMatchingSnapshot();
});
