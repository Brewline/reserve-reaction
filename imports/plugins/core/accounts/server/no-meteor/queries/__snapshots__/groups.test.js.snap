// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws access-denied if not allowed 1`] = `"User does not have permissions to view groups"`;
